zerver/lib/import_realm.py

```diff
@@ -444,6 +452,53 @@ def fix_message_rendered_content(
             )
 
 
+def valid_near_link_parts(near_link_parts: list[str]) -> bool:
+    return (
+        3 <= len(near_link_parts) <= 7
+        and near_link_parts[0] == "#narrow"
+        and near_link_parts[1] in ["channel", "stream"]
+    )
+
+
+def fix_near_links_in_messages(rendered_content: str) -> str:
+    soup = BeautifulSoup(rendered_content, "html.parser")
+    near_links = soup.find_all(
+        lambda tag: tag.name == "a"
+        and tag.has_attr("href")
+        and ("#narrow/stream/" in tag.get("href") or "#narrow/channel/" in tag.get("href"))
+    )
```

Comment:
You are missing some sort of check that this is for the same Zulip instance -- e.g. them being relative links (starting with `/`. I suppose also a link that's an absolute link to the pre-export server's URL should have this apply as well, but I think our Markdown system already converts all of those to relative via `rewrite_local_links_to_relative`, so we should be OK there. 

Concretely I think this means you want to be checking `startswith("/#narrow/channel/` instead.

Reply:
Ah, `rewrite_local_links_to_relative` includes the "/" as part of the prefix, so when removing it, the resulting relative link doesn't begin with "/".
 
```python
def test_relative_link(self) -> None:
        realm = get_realm("zulip")
        sender_user_profile = self.example_user("othello")
        message = Message(sender=sender_user_profile, sending_client=get_client("test"))

        msg = "http://zulip.testserver/#narrow/channel/999-hello"
        self.assertEqual(
            markdown_convert(msg, message_realm=realm, message=message).rendered_content,
            '<p><a href="#narrow/channel/999-hello">http://zulip.testserver/#narrow/channel/999-hello</a></p>',
        )

        msg = f"http://zulip.testserver/user_uploads/{realm.id}/ff/file.txt"
        self.assertEqual(
            markdown_convert(msg, message_realm=realm, message=message).rendered_content,
            f'<p><a href="user_uploads/{realm.id}/ff/file.txt">http://zulip.testserver/user_uploads/{realm.id}/ff/file.txt</a></p>',
        )
      ...
```

Reply:
I included a fix for this in this prep commit, [markdown: Fix relative link reformatting.](https://github.com/zulip/zulip/pull/32272/commits/6c723c2c183de4cc3b90763f4aff75148963bd2f)

Reply:
EDIT: this commit: 598272b27f0711bf32f0dfe2714a6bf09554fb26

Reply:
Since previous Markdown renderings are in the database, I think you just need to support both `"/#narrow"` and `/#narrow` for this logic.

Meanwhile, `markdown: Fix relative link reformatting.` is a high-risk (but probably correct) change, because of how annoying it is to fix historical bugs in those links. Can you extract it to its own pull request, and post in "#api design" about it? I suspect we'll want to do your fix, but with a documented API change, and some hand-testing that mobile and terminal clients don't break.